### PR TITLE
[4.0] Fix SCSS build script

### PR DESCRIPTION
**Pull request by bembelimen**
Pull Request for Issue #24144 .

### Summary of Changes
- [x] Correct the SCSS build paths for windows systems
- [x] Correct the wrong "_" (underscore) test regexp
- [x] Enforce correct sequence for SCSS compiling

### Testing Instructions
- Delete "media" folder
- Use "npm i" as before => now all scss files are at the correct position

### CODE REVIEW

**Comment by bembelimen** [comment]
@dgrammatiko perhaps you have some seconds to look, if you agree with the changes? Thanks!

**Comment by dgrammatiko** [comment]
@bembelimen I think is better to continue with a then, like:
```js
  // Loop to get the files that should be compiled via parameter
  folders.forEach((folder) => {
    Recurs(folder, ['*.js', '*.map', '*.svg', '*.png', '*.gif', '*.swf', '*.html', '*.json']).then(
      (filesRc) => {
        filesRc.forEach(
          (file) => {
            // Don't take files with "_" but "file" has the full path, so check via match
            if (file.match(/\.scss$/) && !file.match(/(\/|\\)_[^\/\\]+$/)) {
                files.push(file);
            }
            if (file.match(/\.css/)) {
              // CSS file, we will copy the file and then minify it in place
              // Ensure that the directories exist or create them
              MakeDir.run(Path.dirname(file).replace('/build/media_source/', '/media/').replace('\\build\\media_source\\', '\\media\\'));
              Fs.copyFileSync(file, file.replace('/build/media_source/', '/media/').replace('\\build\\media_source\\', '\\media\\'));
              Fs.writeFileSync(
                file.replace('/build/media_source/', '/media/').replace('\\build\\media_source\\', '\\media\\').replace('.css', '.min.css'),
                UglyCss.processFiles([file], { expandVars: false }),
                { encoding: 'utf8' },
              );

              // eslint-disable-next-line no-console
              console.log(`CSS file copied/minified: ${file}`);
            }
          },
          (error) => {
            // eslint-disable-next-line no-console
            console.error(`something exploded ${error}`);
          },
        );
            
            return files;
      },
    ).then(theFiles => {
        theFiles.forEach((inputFile) => {
            CompileScss.compile(inputFile, options);
        });
    });
  });
```

Also why the heck are there still GIF files in the media folder? It's 2019...

**Comment by bembelimen** [comment]
> @bembelimen I think is better to continue with a then, like:

Thanks, that's exactly the feedback I needed!

> Also why the heck are there still GIF files in the media folder? It's 2019...

https://github.com/joomla/joomla-cms/tree/4.0-dev/build/media_source/mod_languages/images

File: build/build-modules-js/compilecss.es6.js [hunk 1]
```diff
@@ -86,11 +87,15 @@ module.exports.compile = (options, path) => {
               },
             );
           },
+        ).then(
+          () => {
```
Comment (dgrammatiko) [inline build/build-modules-js/compilecss.es6.js]: This probably won't work, the previous 'then' needs to return the array of the files and this one needs to catch it: 
so something like this in this line: `(someVar) => {`
and something like `return files;` around 87

**Comment by brianteeman** [comment]
From my logs after running this
```

Transpiling ES6 file: C:\htdocs\joomla-cms\build\media_source\system\js\fields\passwordview.es6.js
Transpiling ES6 file: C:\htdocs\joomla-cms\build\media_source\system\js\fields\passwordview.es6.js
Transpiling ES6 file: C:\htdocs\joomla-cms\build\media_source\system\js\fields\passwordview.es6.js
Transpiling ES6 file: C:\htdocs\joomla-cms\build\media_source\system\js\fields\passwordview.es6.js
```
this was what happened for all es6

And on my windows system the paths were being written differently depending on the location

```
SCSS File compiled: C:\htdocs\joomla-cms\media\plg_installer_webinstaller\css\client.css
SCSS File compiled: C:\htdocs\joomla-cms/installation/template/css/template.css
```

**Comment by dgrammatiko** [comment]
@brianteeman it's fine

**Comment by wilsonge** [comment]
I have tested this item :white_check_mark: successfully on 366f8c3dab831a77c1457994e297735879fb1749<br /><br />Before patch the following files were compiled

```
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/templates/cassiopeia/css/offline.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/bootstrap.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/template.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/font-awesome.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/template-rtl.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/templates/cassiopeia/css/template-rtl.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/templates/cassiopeia/css/template.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/installation/template/css/template.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/installation/template/css/template-rtl.css
```

After patch:

```
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/templates/cassiopeia/css/offline.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/font-awesome.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/template.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/bootstrap.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/administrator/templates/atum/css/template-rtl.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/templates/cassiopeia/css/template.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/media/system/css/fields/joomla-field-simple-color.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/templates/cassiopeia/css/template-rtl.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/media/system/css/fields/joomla-field-switcher.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/media/system/css/fields/joomla-field-permissions.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/media/system/css/fields/joomla-field-media.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/media/plg_installer_webinstaller/css/client.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/installation/template/css/template.css
SCSS File compiled: /Users/georgewilson/workspace/opensource/joomla-cms/installation/template/css/template-rtl.css
```

So it seemed to find a bunch of extra files so I think this patch works on OSX<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24169">issues.joomla.org/tracker/joomla-cms/24169</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**